Add EiffelCause and $JOB_URL/eiffel/build API endpoint

magnusbaeck opened the pull request:
A new EiffelCause cause containing Eiffel event links is recognized the queue listener and populates the ActT event.
    
The new $JOB_URL/eiffel/build API endpoint complements the regular /build endpoint but also recognizes an "eiffellinks" key that, specified, gets parsed and ends up in the EiffelCause passed to started build. This allows clients to add e.g. CONTEXT and CAUSE to the Jenkins-originating activities.

There's one known bug/limitation: The EiffelCause class doesn't `@Export` list of links, only the short description. This shouldn't be a problem since the equivalent information is available from the events stored in the EiffelActivityAction that's also connected    to the run.

Closes #9.

GLundh commented:
This will take a few days. Quite swamped and it is not a "super small" PR.

GLundh commented on src/main/java/com/axis/jenkins/plugins/eiffel/eiffelbroadcaster/BuildWithEiffelLinksAction.java:
```diff
@@ -0,0 +1,279 @@
+     */
+    public static final String URL_SUFFIX = "eiffel";
+
+    private JobT job;
```
final?

magnusbaeck replied:
Yes, done.

GLundh commented on src/main/java/com/axis/jenkins/plugins/eiffel/eiffelbroadcaster/BuildWithEiffelLinksAction.java:
```diff
@@ -0,0 +1,279 @@
+        }
+
+        List<Action> actions = new ArrayList<>();
+        JSONObject formData = null;
```
Objects are already initialized as null

magnusbaeck replied:
Done.

GLundh commented on src/main/java/com/axis/jenkins/plugins/eiffel/eiffelbroadcaster/BuildWithEiffelLinksAction.java:
```diff
@@ -0,0 +1,279 @@
+                            "Missing or invalid contents of \"json\" form field: " + e.toString(), e));
+        }
+        try {
+            ParametersDefinitionProperty pp = (ParametersDefinitionProperty) job.getProperty(
```
Not needed. You are already defining the class in the method call.

magnusbaeck replied:
Indeed. I could've sworn I had problems with an unexpected need for a cast but it must've been in some other place.

GLundh reviewed (commented):
The code and tests LGTM. Some tiny nits. Will this work with the rebuilder-plugin?

magnusbaeck commented on src/main/java/com/axis/jenkins/plugins/eiffel/eiffelbroadcaster/BuildWithEiffelLinksAction.java:
```diff
@@ -0,0 +1,279 @@
+     * and only slightly modified.
+     */
+    public Cause getCallerCause(StaplerRequest req) {
+        Cause cause;
```
Indeed not. Leftover from the original code but became unused when I introduced multiple return statements.

magnusbaeck commented:
I've addressed your comments @GLundh (and removed an unused method argument).

> Will this work with the rebuilder-plugin?

Define "work" :-) Ideally this should be modeled in Eiffel with a "retrigger" link to the ActT of the build that's being rebuilt. At the moment this isn't possible because there's no "retrigger" link type but it has been discussed in https://github.com/eiffel-community/eiffel/issues/246. What's going to happen now is that we'll set up a CAUSE link to the build that's being rebuilt (since [BuildCause](https://github.com/jenkinsci/rebuild-plugin/blob/master/src/main/java/com/sonyericsson/rebuild/RebuildCause.java) extends Cause.UpstreamCause, which we already deal with) but I'll verify that to make sure. That's isn't a perfect reflection of reality since it'll look like a downstream build of itself but we'll at least not break the Eiffel DAG when we rebuild something. The only other option with the current version of the protocol would be to copy the rebuilt build's links but that's not perfect either.

magnusbaeck commented:
This is what a rebuilt build looks like:
```
{
  "data": {
    "name": "test",
    "triggers": [
      {
        "type": "OTHER",
        "description": "Started by user anonymous"
      },
      {
        "type": "EIFFEL_EVENT",
        "description": "Rebuilds build #60"
      }
    ]
  },
  "links": [
    {
      "type": "CAUSE",
      "target": "24b6bcc8-0724-4929-b9a5-f8ef1cc85c44"
    }
  ],
  "meta": {
    "type": "EiffelActivityTriggeredEvent",
    "version": "1.1.0",
    "id": "5e63f3fe-c458-45c4-ae07-60594a744f19",
    "source": {
      "name": "JENKINS_EIFFEL_BROADCASTER"
    },
    "time": 1614084282717
  }
}
```
And again, the target of the CAUSE link (24b6bcc8-0724-4929-b9a5-f8ef1cc85c44) is the ActT event of build \#60.

GLundh commented:
Ok. I think it's reasonable. But please know that if we want to manage the rebuild in other ways, there is an @Extension point in rebuilder that allows us to disable the rebuilder actions on certain builds, so we instead can construct a new "rebuild"-link of our own.
